app/Models/OAuth/Token.php

```diff
@@ -43,6 +70,58 @@ public function revoke()
         return $saved;
     }
 
+    public function scopeValidAt($query, $time)
+    {
+        return $query->where('revoked', false)->where('expires_at', '>', $time);
+    }
+
+    public function validate()
+    {
+        if (empty($this->scopes)) {
+            throw new InvalidScopeException('Tokens without scopes are not valid.');
+        }
+
+        if ($this->client === null) {
+            throw new InvalidScopeException('The client is not authorized.', 'unauthorized_client');
+        }
+
+        if ($this->isClientCredentials() && in_array('*', $this->scopes, true)) {
+            throw new InvalidScopeException('* is not allowed with Client Credentials');
+        }
+
+        if (in_array('chat.write', $this->scopes, true)) {
+            // only clients owned by bots can request this scope.
+            if (!$this->client->user->isBot()) {
+                throw new InvalidScopeException('This scope is only available for chat bots.');
+            }
+
+            // in the case of client credentials, delegation must be enabled on the token.
+            if ($this->isClientCredentials() && !$this->canDeletage()) {
+                throw new InvalidScopeException('bot scope required.');
+            }
+        }
+
+        if ($this->canDeletage()) {
+            if (!$this->client->user->isBot()) {
+                throw new InvalidScopeException('This scope is only available for chat bots.');
+            }
+
+            if (!$this->isClientCredentials()) {
+                throw new InvalidScopeException('bot scope is only valid for client_credentials tokens.');
+            }
```

Comment:
I wonder if it's worth rewriting these conditionals to group the `isClientCredentials()` stuff together - might make it more obvious which restrictions apply to which auth type?

Reply:
These are grouped by scope/permission requested, grouping by `isClientCredentials()` just means more scope and user bot checks need to be added or duplicated to the conditional since the token type is the exception not the rule ("you can do `a` unless in case of `x`", vs. "in case of `x` you can't do `a`...unless `y` but not if `c`")